contracts/BonusRewards.sol

```diff
@@ -3,11 +3,369 @@
+        require(bonus.bonusTokenAddr == _token, "BonusRewards: wrong pool");
+        require(bonus.endTime + WEEK < block.timestamp, "BonusRewards: not ready");
+        balance = bonus.remBonus;
+        pools[_lpToken].bonuses[_poolBonusId].remBonus = 0;
+      }
+
+      IERC20(_token).transfer(owner(), balance);
+    }
+  }
+
+  function setResponders(address[] calldata _responders) external override onlyOwner {
+    responders = _responders;
+  }
+
+  function setPaused(bool _paused) external override {
+    require(_isAuthorized(msg.sender, responders), "BonusRewards: caller not responder");
+    paused = _paused;
+  }
+
+  function getPool(address _lpToken) external view override returns (Pool memory) {
+    return pools[_lpToken];
+  }
+
+  function getUser(address _lpToken, address _account) external view override returns (User memory, uint256[] memory) {
+    return (users[_lpToken][_account], viewRewards(_lpToken, _account));
+  }
+
+  function getAuthorizers(address _lpToken, address _bonusTokenAddr) external view override returns (address[] memory) {
+    return allowedTokenAuthorizers[_lpToken][_bonusTokenAddr];
+  }
+
+  function getResponders() external view override returns (address[] memory) {
+    return responders;
+  }
+
+  function viewRewards(address _lpToken, address _user) public view override returns (uint256[] memory) {
+    Pool memory pool = pools[_lpToken];
+    User memory user = users[_lpToken][_user];
+    uint256[] memory rewards = new uint256[](pool.bonuses.length);
+    if (user.amount <= 0) return rewards;
+
+    uint256 rewardsWriteoffsLen = user.rewardsWriteoffs.length;
+    for (uint256 i = 0; i < rewards.length; i ++) {
+      Bonus memory bonus = pool.bonuses[i];
+      if (bonus.startTime < block.timestamp && bonus.remBonus > 0) {
+        uint256 lpTotal = IERC20(_lpToken).balanceOf(address(this));
+        uint256 bonusForTime = _calRewardsForTime(bonus, pool.lastUpdatedAt);
+        uint256 bonusPerToken = bonus.accRewardsPerToken + bonusForTime / lpTotal;
+        uint256 rewardsWriteoff = rewardsWriteoffsLen <= i ? 0 : user.rewardsWriteoffs[i];
+        uint256 reward = user.amount * bonusPerToken / CAL_MULTIPLIER - rewardsWriteoff;
+        rewards[i] = reward < bonus.remBonus ? reward : bonus.remBonus;
+      }
+    }
+    return rewards;
+  }
+
+
+  function getPoolList() external view override returns (address[] memory) {
+    return poolList;
+  }
+
+  /// @notice update pool's bonus per staked token till current block timestamp, do nothing if pool does not exist
+  function _updatePool(address _lpToken) private {
+    Pool storage pool = pools[_lpToken];
+    uint256 poolLastUpdatedAt = pool.lastUpdatedAt;
+    if (poolLastUpdatedAt == 0 || block.timestamp <= poolLastUpdatedAt) return;
+    pool.lastUpdatedAt = block.timestamp;
+    uint256 lpTotal = IERC20(_lpToken).balanceOf(address(this));
+    if (lpTotal == 0) return;
+
```

Comment:
the pool needs to be update before. Cause it will be used to calculate the accu rewards for the first depositor. If not, the first deposit will get a little more rewards than it should.